### DISCUSSION TITLE
Synchronous IAsyncEnumerable iterator method

### DISCUSSION THREAD

**Discussion by verdie-g** [dotnet/csharplang#4952] [Language Ideas]
Hello,

When trying to write a synchronous iterator method returning an IAsyncEnumerable
```csharp
async IAsyncEnumerable<int> Enumerate()
{
    foreach (int i in new[] { 1, 2, 3 }) yield return i;
}
```
a warning is raised
> [CS1998] This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.

and an error is raised by removing the `async` keyword
> [CS8403] Method with an iterator block must be 'async' to return 'IAsyncEnumerable<int>'

One could say use `IEnumerable` but I'm implementing an interface here.

Is there any blocker to allow non-async iterator methods? For example:
```csharp
IAsyncEnumerable<int> Enumerate()
{
    foreach (int i in new[] { 1, 2, 3 }) yield return i;
}
```

**Comment by CyrusNajmabadi** [comment]
@jcouv ?

**Reply by jcouv** [reply]
This type of adaptation can already be solved via helper libraries (see [IAsyncEnumerable<TSource> ToAsyncEnumerable<TSource>(IEnumerable<TSource>)](https://github.com/dotnet/reactive/blob/main/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToAsyncEnumerable.cs)).

**Reply by verdie-g** [reply]
It can, but it sounds very cumbersome to reference a library just for that ToAsyncEnumerable method.

I guess it depends how frequent my use-case is. I personally have this issue in most of recent projects when I need to mock an interface that returns an IAsyncEnumerable.

**Reply by CyrusNajmabadi** [reply]
You can also just suppress the warning in this case with a comment explaining that this is a case where you can produce this all synchronously.

**Reply by CyrusNajmabadi** [reply]
or, you can add a dummy `await Task.CompletedTask;` to make things happy :)

**Reply by jnm2** [reply]
I suppress the warning for ordinary `async Task` methods too. It's not a bad thing to have handleable exceptions marshaled into a returned Task instance instead of throwing before the task is even awaited.
